src/js/common/schemaform/ObjectField.jsx

```diff
@@ -0,0 +1,184 @@
+import React from 'react';
+
+import {
+  deepEquals,
+  getDefaultFormState,
+  orderProperties,
+  retrieveSchema,
+  shouldRender,
+  getDefaultRegistry,
+  setState
+} from 'react-jsonschema-form/lib/utils';
+
+/*
+ * This is largely copied from the react-jsonschema-form library,
+ * but with the way descriptions are used changed
+ */
+
+function objectKeysHaveChanged(formData, state) {
+  // for performance, first check for lengths
+  const newKeys = Object.keys(formData);
+  const oldKeys = Object.keys(state);
+  if (newKeys.length < oldKeys.length) {
+    return true;
+  }
+  // deep check on sorted keys
+  if (!deepEquals(newKeys.sort(), oldKeys.sort())) {
+    return true;
+  }
+  return false;
+}
+
+class ObjectField extends React.Component {
+  static defaultProps = {
+    uiSchema: {},
+    errorSchema: {},
+    idSchema: {},
+    registry: getDefaultRegistry(),
+    required: false,
+    disabled: false,
+    readonly: false,
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = this.getStateFromProps(props);
+    this.onPropertyChange = this.onPropertyChange.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const state = this.getStateFromProps(nextProps);
+    const { formData } = nextProps;
+    if (formData && objectKeysHaveChanged(formData, this.state)) {
+      // We *need* to replace state entirely here has we have received formData
+      // holding different keys (so with some removed).
+      this.state = state;
+      this.forceUpdate();
+    } else {
+      this.setState(state);
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return shouldRender(this, nextProps, nextState);
+  }
+
+  onPropertyChange(name) {
+    return (value, options) => {
+      this.asyncSetState({ [name]: value }, options);
+    };
+  }
+
+  getStateFromProps(props) {
+    const { schema, formData, registry } = props;
+    return getDefaultFormState(schema, formData, registry.definitions) || {};
+  }
+
+  isRequired(name) {
+    const schema = this.props.schema;
+    return Array.isArray(schema.required) &&
+      schema.required.indexOf(name) !== -1;
+  }
+
+  asyncSetState(state, options = { validate: false }) {
+    setState(this, state, () => {
+      this.props.onChange(this.state, options);
+    });
+  }
+
+  render() {
+    const {
+      uiSchema,
+      errorSchema,
+      idSchema,
+      name,
+      required,
+      disabled,
+      readonly,
+      onBlur
+    } = this.props;
+    const { definitions, fields, formContext } = this.props.registry;
+    const { SchemaField, TitleField } = fields;
+    const schema = retrieveSchema(this.props.schema, definitions);
+
+    // description and title setup
```

Comment:
That makes sense. For now can we document the changes we've made somewhere so we can keep track of those in case the first 2 PRs ever get merged?